webrender/src/platform/unix/font.rs

```diff
@@ -228,100 +262,107 @@ impl FontContext {
+        let top = unsafe { (*slot).bitmap_top };
+        let left = unsafe { (*slot).bitmap_left };
+        let mut final_buffer = vec![0; (actual_width * actual_height * 4) as usize];
+
+        // Extract the final glyph from FT format into RGBA8 format, which is
+        // what WR expects.
+        for y in 0..actual_height {
+            // Get pointer to the bytes for this row.
+            let mut src = unsafe {
+                bitmap.buffer.offset((y * bitmap.pitch) as isize)
+            };
+
+            for x in 0..actual_width {
+                let value = match pixel_mode {
+                    FT_Pixel_Mode::FT_PIXEL_MODE_MONO => {
+                        let mask = 0x80 >> (x & 0x7);
+                        let byte = unsafe { *src.offset((x >> 3) as isize) };
                         let alpha = if byte & mask != 0 { 0xff } else { 0 };
-                        final_buffer.extend_from_slice(&[0xff, 0xff, 0xff, alpha]);
+                        [0xff, 0xff, 0xff, alpha]
                     }
-                }
-                FT_Pixel_Mode::FT_PIXEL_MODE_GRAY => {
-                    for x in dst_start .. dst_end {
-                        let alpha = unsafe {
-                            *base.offset((x + offset_x) as isize)
-                        };
-                        final_buffer.extend_from_slice(&[0xff, 0xff, 0xff, alpha]);
+                    FT_Pixel_Mode::FT_PIXEL_MODE_GRAY => {
+                        let alpha = unsafe { *src };
+                        src = unsafe { src.offset(1) };
+                        [0xff, 0xff, 0xff, alpha]
                     }
-                }
-                FT_Pixel_Mode::FT_PIXEL_MODE_LCD => {
-                    for x in dst_start .. dst_end {
-                        let src_x = ((x + offset_x) * 3) as isize;
-                        assert!(src_x+2 < bitmap.pitch as isize);
-                        let t = unsafe {
-                            slice::from_raw_parts(base.offset(src_x), 3)
-                        };
-                        final_buffer.extend_from_slice(&[t[2], t[1], t[0], 0xff]);
+                    FT_Pixel_Mode::FT_PIXEL_MODE_LCD => {
+                        let t = unsafe { slice::from_raw_parts(src, 3) };
+                        src = unsafe { src.offset(3) };
+                        [t[2], t[1], t[0], 0xff]
                     }
-                }
-                _ => panic!("Unsupported {:?}", pixel_mode)
-            }
-            for _x in dst_end .. dimensions.width as i32 {
-                final_buffer.extend_from_slice(&[0xff, 0xff, 0xff, 0]);
+                    _ => panic!("Unsupported {:?}", pixel_mode)
+                };
+                let i = 4 * (y * actual_width + x) as usize;
+                let dest = &mut final_buffer[i..i+4];
+                dest.clone_from_slice(&value);
             }
-            assert_eq!(final_buffer.len(), ((y+1) * dimensions.width * 4) as usize);
```

Comment:
we need an assertion check here to ensure that `src` reached exactly the end of the source buffer